### ISSUE TITLE
Need zoom feature and Download vector graphics from server feature

### ISSUE THREAD

**Issue by anishkumar11** [tarek360/RichPath#45]
Hi Tarek,

I like your library. I have very useful and powerful. But two features are missing in your library.
1) Zoom feature. It will help to fill the path tiny places.
2) If we have put graphic xml file on the server or in the sdcard. then it should be load by your library.

**Comment by tarek360** [comment]
I like your suggestions, I will be more than happy if you created a pull
request for each feature.

**Comment by anishkumar11** [comment]
Hi Tarek. I happy you replied about my query.

**Comment by anishkumar11** [comment]
I have created the pull requests

**Comment by AmineLAHRIM** [comment]
please how about the zoom ?

**Comment by anishkumar11** [comment]
Zoom functionality like we have in photo library. And we can change path
color in zoom mode too.

**Comment by tarek360** [comment]
Hi,
@anishkumar11 where is that pull request?

@AmineLAHRIM could you explain more about your use case?
do you need to zoom at the level of the path or the view?

**Comment by anishkumar11** [comment]
Yes i need zoon path level too. So we can event this path is clicked in
zoom mode.  I tried to implement this but fail.

**Comment by AmineLAHRIM** [comment]
just in view i try with TouchImageView.java and extends from RichPathView the Zoom not working but if the class extends from the SVGImageView that from the library of AndroidSVG it works but the problem that i need to work on your library because of having some great functionality and thank you so much Tarek. for your reply 

TouchImageView.java =>
https://gist.github.com/Antarix/65cfc180c05524ce616f


the right problem is in RichPathDrawale because if i make set imageDrawable(A_SIMPLE_DRAWABLE) the zoom work but if I make imageDrawable(richPathDrawable) the zoom didn't work because of extendes methods like draw() ononBoundsChange() ...

but finally we need that richPathDrawable because is part of the library.

**Comment by AmineLAHRIM** [comment]
the right problem is in RichPathDrawale because if i make set imageDrawable(A_SIMPLE_DRAWABLE) the zoom work but if I make imageDrawable(richPathDrawable) the zoom didn't work because of extendes methods like draw() ononBoundsChange() ...

**Comment by anishkumar11** [comment]
Yes correct these problems were arise when i try to modifed your library.

**Comment by anishkumar11** [comment]
sI

**Comment by tarek360** [comment]
@AmineLAHRIM cool, you have figured where the problem with a zoom view, are you able to fix that in RichPathDrawabl? What exactly is going wrong in RichPathDrawabl?

**Comment by AmineLAHRIM** [comment]
@tarek360 we cloud not fix it i try so hard about everyday with no result could you please help us

**Comment by tarek360** [comment]
@AmineLAHRIM I checked it and I found that [TouchImageView.java](https://gist.github.com/Antarix/65cfc180c05524ce616f
) is an ImageView but in your case, you need a ViewGroup, you need a layout to wrap the RichPathView to be able to zoom.

There are many many samples on the internet for zoom layout, I found one of them [here](https://github.com/shopgun/zoomlayout)

here is an example I tested and the zoom works:

```xml
    <com.shopgun.android.zoomlayout.ZoomLayout
        android:layout_width="200dp"
        android:layout_height="200dp">

        <com.richpath.RichPathView
            android:id="@+id/ic_command"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            app:vector="@drawable/ic_command" />

    </com.shopgun.android.zoomlayout.ZoomLayout>
```

![zoom](https://user-images.githubusercontent.com/8615768/73585700-420c1680-44d6-11ea-8fbe-748f0032049a.gif)

**Comment by tarek360** [comment]
@AmineLAHRIM I checked it and I found that [TouchImageView.java](https://gist.github.com/Antarix/65cfc180c05524ce616f
) is an ImageView but in your case, you need a ViewGroup, you need a layout to wrap the RichPathView to be able to zoom.

There are many many samples on the internet for zoom layout, I found one of them [here](https://github.com/shopgun/zoomlayout)

here is an example I tested and the zoom works:

```xml
<com.shopgun.android.zoomlayout.ZoomLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:id="@+id/touchImageView"
    android:layout_width="match_parent"
    android:layout_height="match_parent">

    <com.richpath.RichPathView
        android:id="@+id/ic_command"
        android:layout_width="200dp"
        android:layout_height="200dp"
        android:layout_margin="4dp"
        android:padding="8dp"
        android:layout_gravity="center"
        app:vector="@drawable/ic_command" />

</com.shopgun.android.zoomlayout.ZoomLayout>
```

![zoom](https://user-images.githubusercontent.com/8615768/73585700-420c1680-44d6-11ea-8fbe-748f0032049a.gif)

**Comment by tarek360** [comment]
@AmineLAHRIM I checked it and I found that [TouchImageView.java](https://gist.github.com/Antarix/65cfc180c05524ce616f
) is an ImageView but in your case, you need a ViewGroup, you need a layout to wrap the RichPathView to be able to zoom.

There are many many samples on the internet for zoom layout,  [I found one of them here](https://github.com/shopgun/zoomlayout)

here is an example I tested and the zoom works:

```xml
<com.shopgun.android.zoomlayout.ZoomLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:id="@+id/touchImageView"
    android:layout_width="match_parent"
    android:layout_height="match_parent">

    <com.richpath.RichPathView
        android:id="@+id/ic_command"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:layout_gravity="center"
        app:vector="@drawable/ic_command" />

</com.shopgun.android.zoomlayout.ZoomLayout>
```

![zoom](https://user-images.githubusercontent.com/8615768/73585700-420c1680-44d6-11ea-8fbe-748f0032049a.gif)

**Comment by AmineLAHRIM** [comment]
Oh my god Thank you for your simple solution that I never think about it.

**Comment by AmineLAHRIM** [comment]
but there is a single problem when you zoom you can't apply any click  listener effect in RichPathView: 
ic_command.setOnPathClickListener(new RichPath.OnPathClickListener() {
            @Override
            public void onClick(RichPath richPath) {
                Log.d("TAG", "getFillColor: " + richPath.getFillColor());
            
                    final int random = new Random().nextInt((max - min) + 1) + min;
                    RichPathAnimator.animate(richPath)
                            .fillColor(random)
                            .start();
            }
        });

**Comment by AmineLAHRIM** [comment]
the only result that i found is to make listener in the zoomlayout on touchevent if there is two fingers zoomLayout.setAllowZoom(true); if not
then zoomLayout.setAllowZoom(false);
then everyting will work fin thank you sir for your help

**Comment by AmineLAHRIM** [comment]
i hight recommend to download the zoomlayout source code and edit it like the following instructions:
1/ declare a public richpathview image in ZoomLayout class
2/ relate the richpathview that you have in MainActivity to it like that
zoomLayout.richpathview = richpathview;
3/ edit onTouchEvent
    public RichPathView richpathview;
    private PointF firstPointerDown = new PointF();
    private boolean isImageMoved = false;

    @Override
    public boolean onTouchEvent(MotionEvent ev) {

        mArray[0] = ev.getX();
        mArray[1] = ev.getY(); 
        PointF curr = new PointF(ev.getX(), ev.getY());

        final int action = ev.getAction() & MotionEvent.ACTION_MASK;
        //final int pointerCount = ev.getPointerCount();

        float deltaXBeforePointerUp;
        float deltaYBeforePointerUp;

        dispatchOnTouch(action, ev);

        switch (action & MotionEvent.ACTION_MASK) {
            case MotionEvent.ACTION_POINTER_DOWN:
                Log.d(TAG, "onTouchEvent: ACTION_POINTER_DOWN DISPATCH");

                break;
            case MotionEvent.ACTION_DOWN:
                Log.d(TAG, "onTouchEvent: ACTION_DOWN DISPATCH");
                firstPointerDown.set(curr);
                Log.d(TAG, "onTouchEvent2: " + curr.x + "  ___ " + firstPointerDown.x);

                break;
            case MotionEvent.ACTION_UP:

                deltaXBeforePointerUp = curr.x - firstPointerDown.x;
                deltaYBeforePointerUp = curr.y - firstPointerDown.y;
                if (Math.abs(deltaXBeforePointerUp) > 0 || Math.abs(deltaYBeforePointerUp) > 0) {
                    Log.d(TAG, "onTouchEvent: ACTION_UP NO_DISPATCH");
                } else {
                    Log.d(TAG, "onTouchEvent: ACTION_UP DISPATCH");
                    if (!isImageMoved) {
                        dispatchEventToRichPathView(ev);
                    }
                }

                //init
                isImageMoved=false;
                break;
            case MotionEvent.ACTION_POINTER_UP:
                break;
            case MotionEvent.ACTION_MOVE:
                deltaXBeforePointerUp = curr.x - firstPointerDown.x;
                deltaYBeforePointerUp = curr.y - firstPointerDown.y;
                if (Math.abs(deltaXBeforePointerUp) > 0 || Math.abs(deltaYBeforePointerUp) > 0) {
                    Log.d(TAG, "onTouchEvent: ACTION_MOVE NO_DISPATCH");
                    isImageMoved=true;
                }
                break;
        }

        scaledPointsToScreenPoints(mArray);
        ev.setLocation(mArray[0], mArray[1]);

        if (!mAllowZoom) {
            return false;
        }


        boolean consumed = mScaleDetector.onTouchEvent(ev);
        consumed = mGestureDetector.onTouchEvent(ev) || consumed;
        if (action == MotionEvent.ACTION_UP) {
            // manually call up
            consumed = mGestureListener.onUp(ev) || consumed;
        }
        return consumed;
    }

    private void dispatchEventToRichPathView(MotionEvent ev) {
        if (this.richpathview != null) {
            richpathview.dispatchTouchEvent(ev);
        }
    }

**Comment by tarek360** [comment]
So it works now?

**Comment by AmineLAHRIM** [comment]
Yeap with two-finger zoom and one finger drag and one-click coloring.

**Comment by AmineLAHRIM** [comment]
special thanks to you tarek.

**Comment by tarek360** [comment]
Are you doing a coloring book app?

**Comment by AmineLAHRIM** [comment]
Yes am trying 😃

**Comment by ManuelCoral1998** [comment]
Remembering what @tarek360  said about the ZoomLayout, I found that this <a href="https://github.com/natario1/ZoomLayout" target="_top">ZoomLayout</a> works fine and allows images to still have their click listener in the RichPathView as long as it has hasClickableChildren set to true.
